clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -296,9 +301,74 @@ impl FunctionType {
         }
     }
 
+    /// Returns the type of `value`, after converting any contract principal
+    /// types to callable types. In an initial transaction, arguments are typed
+    /// as contract principals, but they must be principal literals, so they
+    /// may be used to call into a contract.
+    pub fn principal_to_callable_type(&self, value: &Value, depth: u8) -> TypeResult {
+        if depth > MAX_TYPE_DEPTH {
+            return Err(CheckErrors::TypeSignatureTooDeep.into());
+        }
+
+        Ok(match value {
+            Value::Principal(PrincipalData::Contract(contract_identifier)) => {
+                TypeSignature::CallableType(CallableSubtype::Principal(contract_identifier.clone()))
+            }
+            Value::Optional(OptionalData {
+                data: Some(inner_value),
+            }) => {
+                TypeSignature::new_option(self.principal_to_callable_type(inner_value, depth + 1)?)?
+            }
+            Value::Response(ResponseData { committed, data }) => {
+                let (ok_type, err_type) = if *committed {
+                    (
+                        self.principal_to_callable_type(data, depth + 1)?,
+                        TypeSignature::NoType,
+                    )
+                } else {
+                    (
+                        TypeSignature::NoType,
+                        self.principal_to_callable_type(data, depth + 1)?,
+                    )
+                };
+                TypeSignature::new_response(ok_type, err_type)?
+            }
+            Value::Sequence(SequenceData::List(ListData {
+                data,
+                type_signature: _,
+            })) => {
+                let inner_type = match data.first() {
+                    Some(inner_val) => self.principal_to_callable_type(inner_val, depth + 1)?,
+                    None => TypeSignature::NoType,
+                };
+                TypeSignature::SequenceType(SequenceSubtype::ListType(ListTypeData::new_list(
+                    inner_type,
+                    data.len() as u32,
+                )?))
+            }
+            Value::Tuple(TupleData {
+                type_signature: _,
+                data_map,
+            }) => {
+                let mut type_map = BTreeMap::new();
+                for (name, field_value) in data_map {
+                    type_map.insert(
+                        name.clone(),
+                        self.principal_to_callable_type(field_value, depth + 1)?,
+                    );
+                }
+                TypeSignature::TupleType(TupleTypeSignature::try_from(type_map)?)
+            }
+            _ => TypeSignature::type_of(value),
+        })
+    }
+
+    /// This method is only used by StacksChainState::can_include_tx. The
+    /// cost of evaluating these type checks are not tracked.
```

Comment:
Can you also add a `WARNING` here that this is _not_ consensus-critical code, and _should never_ be called from consensus-critical code?  Thanks!

Reply:
Added in 7cd1293737682ff35125f2b1069cb7cbce408ff5